c10/core/TensorImpl.cpp

```diff
@@ -60,14 +60,26 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
       storage_offset_(0),
       numel_(0),
       data_type_(data_type),
-      device_opt_(device_opt),
-      key_set_(key_set) {
+      device_opt_(device_opt) {
   if (!key_set.empty()) {
     AT_ASSERT(data_type.id() ==  caffe2::TypeIdentifier::uninitialized() ||
               device_opt_.has_value());
     // UndefinedTensorImpl is a singleton, so we skip logging it
     C10_LOG_API_USAGE_ONCE("tensor.create");
   }
+  // After we removed Autograd keys from globally enabled set, every Tensor must be created with
+  // a backend DispatchKey and an AutogradBackend key.
+  // We automatically add the corresponding autograd key to key_set_ so that backends can stay
+  // in the old way of only registering with backend key like DispatchKey::CPU.
+  for (uint8_t i = static_cast<uint8_t>(DispatchKey::CPU); i < static_cast<uint8_t>(DispatchKey::NumBackendKeys); i++) {
```

Comment:
This logic adding autograd keys is a bit hacky. Optionally we can move this to caller sites if it's a better design choice. (either in this PR or a followup)

Reply:
Ah actually the failed cpp tests are related to this logic that we now automatically added AutogradKeys in TensorImpl. But cpp tests were tested without the global `Autograd` key enable set. So where we want to put this logic will affect how we fix the cpp tests. 
Should be an easy fix either way but would like to hear more thoughts on what's better from API design perspective. :D

Reply:
I think it's better to add the autograd key here rather than make call sites do it, but it feels weird to have to do a linear pass through the backend keys, do we really have to? Seems like there's some mask-and-shift fanciness we could do here, but admittedly I haven't thought it through.

Also I'd suggest having a utility function we just call here, `getAutogradKeyForBackend(key_set)` that does the actual work. (I'm basing that on the assumption we'll be doing this elsewhere, but I'm thinking we will, right?)

Reply:
Mask-and-shift would make this code branchless. Mask out everything except (e.g., CPU) bit, and then shift it to its AutogradCPU position, then OR it into the original site.

Reply:
And then you could get rid of `NumBackendKeys`